### PR TITLE
Context-Based: Grab Intent (Bayport)

### PR DESCRIPTION
**Pull request by tigercat2000**
This system almost entirely overhauls grab code, with a bunch of new things.

Original list of features from the original PR:
> **Changes:**
- [x] Grab animations
- [x] New grab icons

>**Features:**
- [x] Headbutting
- [x] Dancing

>### Details

>Feature | Description | Trigger | Required | Target
---|---|---|---|---
Headbutting | Causes 5 damage to the victim and 3 damage to assailant | Click on victim with harm intent | Aggressive grab | ~~Head~~ Torso/Groin
Dancing | Constantly facing each other | Two people grabbing each other passively | n/A | n/A

Note about above table: Crossed out things have been changed for compatibility reasons

TL:DR; New great grab stuff. 

## Screenshot section

>Baypost original: **Grab Transition Animations**
Wow

Demonstration videos are screenshots too, right?
[Demo videoee](http://www.youtube.com/watch?v=SIKF_aiPSdg) *cliiiiickk ittt*

**Dancing**
dancedance

**Neck grab**
on da tin

### CODE REVIEW

**Comment by Spacemanspark** [comment]
Might want to turn halloss into stamina.

**Comment by tigercat2000** [comment]
@Spacemanspark AFAIK stamina affects things differently than halloss

**Comment by TheDZD** [comment]
I do agree with Spaceman about the point of turning halloss into stamina, halloss has largely been replaced with stamina damage here.
The other thing I would probably change is jointbreaking, being able to fracture a bone with a neckgrab can easily be abused to permanently paincrit somebody. I'm not certain how I feel about knifing, muzzling, and pinning, and would like some clarification about how long the latter two last/if you can still break away from said things by using Resist.

**Comment by tigercat2000** [comment]
@TheDZD Resist is unchanged- 25% chance to resist out of an aggressive grab every time you hit resist, 5% chance for a neckgrab.

Muzzling, Blinding, and Pinning last indefinitely during the duration of being grabbed, as it will constantly keep the person mute/blind/weakened while you have that selected, but resisting will still work (AFAIK).

Knifing is kiiiinda pointless, but it looks cool. 

Jointbreaking is a bit off with this, since we don't have bay12's magical joint-dislocating system, so I just changed it to a fracture. It's pretty easy to remove/tweak with a longer time.

**Comment by johannhawk** [comment]
hey there, first github comment here.
and to DZD i just want to say that i can choke a person to death and while they are being choked they cant speak or see well or move around from being unconscious after enough choking.

muzzling and pinning is just a less lethal version of choking so it isnt so bad.

**Comment by Fox-McCloud** [comment]
I don't like a lot of things in this; they're rife for abuse in a non-RP environment:

- Neck fracturing
 - Not only unrealistic, but this combined with pinning, jointlocking, and instant aggressive grabs pretty much means that you can permastun multiple people if they're anywhere near tables-potentially even killing multiple poeple.
- Throat cutting 
 - Just no. 60 oxy will instantly knock them out and 60 brute is insane; two clicks and the person is dead--again just from a neck grab. Changelings don't need even more of a buff.
- Jointlocking
 - Table perma-stunning is bad enough; this takes it to a whole new level
- Pinning
 - See above
- Instant aggressive grabbing
 - Yeah, no. It's already easy to table someone who's paying attention--this will make it impossible to counter.

To be honest, about the only thing I really like here is the dancing, HUD, better ways of handling grabs, and the different pixel-shifts for grabs.

While I'm sure some of these are ok in an RP environment, it's going to generate absolute paranoia amongst a lot of player as they'll not want to be in close proximity to anyone for fear they'll get perma-stunned by a simple grab.

As someone who's a long time Goon player, quick-grabbing and quick-knockouts from grabs are by far the #1 cause of why people are so scared to death to be close to each other.

**Comment by FreeStylaLT** [comment]
so now you will be truly able to shank people

**Comment by johannhawk** [comment]
what about...having a option to pin people instead of choking them to death which already immobilizes them.

**Comment by tigercat2000** [comment]
@Fox-McCloud Pinning is not nearly as powerful as table perma-stunning. You require an aggressive grab, and the slightly poke will disturb it, then they get up a lot faster than being tabled.

Instant aggressive grabbing is an incredibly finicky system that requires both people to be facing the same way, one behind the other, It is pitifully easy to counter by just click-turning. You can't even abuse it while the other person is lying down unless you know exactly what you are doing, since lying sprites are confusing as to the actual DIR to even experienced players.

On your other points, I pretty much agree. On knifing, to a point, but considering that you are basically dead by the point that they have a neckgrab on you...

**Comment by Krausus** [comment]
> Instant aggressive grab
> Grab someone from behind
> Instant aggressive grabbing is an incredibly finicky system

Getting behind someone is 1 or 2 presses of a movement key, depending on whether there's something solid on the other side of them. Also, I'm not aware of anything else in the game which depends on a player's facing direction, and would say that's probably an intentional design decision, since micro-managing the way you're facing sounds awful.

> Attack the victim with an edgy item
> Causes 60 brute (divided into 3*20) and 60 oxyloss

What.

> you are basically dead by the point that they have a neckgrab on you

Then why bother with any of this? If neckgrabs are already deadly, why make them **deadlier**? Not that this is necessarily true anyway - if there's more than 2 people on the server, there's a chance for the victim to get saved, but probably not if they're dead in a matter of seconds.

The cosmetic changes are nice, though.

**Comment by ZomgPonies** [comment]
I think ideally you should put up a forum post with a vote. Perhaps the vote needs to be on an external site so people can vote for more than one option.

**Comment by tigercat2000** [comment]
@Krausus Just being behind someone isn't enough. You have to face the same direction as them, which is albiet just a click in front of them, and it is intentional design decision to micromanage direction facing. Some people might not like worrying about the micromanaging, but really, unless you for some reason never shift-click people (which automatically faces you towards them), it won't be a very big problem, in my opinion.

I felt that the neckgrab already being deadly point needed to be brought up, but if community opinion hates knifing, I'm fine with removing it. It's terrible snowflake code anyways (no offense Hubblenaut)

**Comment by Fox-McCloud** [comment]
It's also work pointing out that it's easy to get someone to face away from you; just make sure you're on harm intent and push up against them; bam you're both facing the same direction and they're instantly primed for an aggressive grab.

It's just not fun gameplay, and the paranoia generated by it will be astounding--not a good thing on a medium RP server, IMO.

**Comment by fj45** [comment]
Zomg, discuss this, don't put it to a damn vote. 


With a few additions maybe this system could be turned into an err... "less then deadly" way to fight? More ruffing someone up then trying to beat their face to a pulp. For barfights and whatnot.

**Comment by ZomgPonies** [comment]
The vote isn't an all or nothing thing. It's to get a better idea of WHAT people think works in it's current form, and what warrants more discussion and tuning.

**Comment by fj45** [comment]
Alright, I can agree with how its being used here, that works ok. Just feel that comments are a much better way to work things out on most things.

**Comment by ZomgPonies** [comment]
The votes happen when after all the comments are said, there is still a deadlock.

**Comment by tigercat2000** [comment]
Going to wait until tomorrow before making any further changes.

**Comment by Krausus** [comment]
Out of curiosity, I decided to see how quickly you could kill someone with a shard of glass...
knifedtodeath
As it turns out, *pretty fast.* You can pin someone before they have time to react, and cover their mouth while you're doing it, so they can't shout for help. Even if knifing weren't a thing, you have ample time to just use some other weapon.

This feels like it progressed straight from "wouldn't it be cool if..." to implementation without stopping to consider whether it was turning grabs into an "I win" button.

**Comment by fj45** [comment]
>This feels like it progressed straight from "wouldn't it be cool if..." to implementation without stopping to consider

Welcome to baycode.

**Comment by Fox-McCloud** [comment]
@fj45 

Yeah, pretty much. It's a realistic feature, yes---but when you factor in non-RP environment, it's just going to get abused literally every round.

**Comment by Bad-Apple797** [comment]
I really like muzzling.
And blinding and dancing are cool I guess.
But the knifing seems silly.
I like headbutting too.

**Comment by Hubblenaut** [comment]
Many things here have been taking into account. I'm aware the system is not balanced at all yet, but the feedback will make its way to the original code branch.
The ability to grab people from behind will be reverted, such will probably knifing.
The pain induced from jointlocking is capped at 40 and is without being used in a clever combination with another technique definitely not OP.
Jointbreaking does not actually break or fracture limbs, however it makes use of the new jointbreak system Bay12 has (which is not entirely finished yet). Afaik there is no surgery needed to fix those.

Also: Resisting out of aggressive and neck grabs is working in baycode and was taken into account for balance, you might want to port it.

**Comment by tigercat2000** [comment]
@Hubblenaut Resisting does work, contrary to what popular belief seems to say.

**Comment by tigercat2000** [comment]
>TMW you realize bay has much higher resist-out-of-grab chance than us and tg

This should help with the appearence that resisting grabs doesn't work. It does, it was just an incredibly unlikely chance.

**Comment by Fox-McCloud** [comment]
Don't really think the upped chance is needed if the other things are just addressed--as it stand, currently only Dancing, the sprites, and headbutting have widespread support.

**Comment by tigercat2000** [comment]
@Fox-McCloud you said yourself that it was very rare to see it working. And, some people do support some of the things that this would help with- namely the behind-grab and muzzling.

**Comment by tigercat2000** [comment]
@TheDZD @Fox-McCloud @ZomgPonies This is pretty much merge-ready, every controversial feature has been commented out, and the framework left as an example.

This basically only consists of the new animations, dancing, and headbutting.

**Comment by Fox-McCloud** [comment]
@tigercat2000 

Is the quick-grabbing that we discussed last night still in-tact?

**Comment by tigercat2000** [comment]
@Fox-McCloud Yes, the last_grab_time set has been removed, I believe

**Comment by Fox-McCloud** [comment]
@tigercat2000 

Aight. Will probably download this and test it out to see how it handles, then I"ll give my full opinion.

File: code/modules/mob/mob_grab.dm [hunk 1]
```diff
@@ -116,87 +215,152 @@
...
+					if(last_hit_zone == "head") //This checks the hitzone the user has selected. In this specific case, they have the head selected.
+						if(affecting.lying)
+							return
+						assailant.visible_message("<span class='danger'>[assailant] thrusts \his head into [affecting]'s skull!</span>") //A visible message for what is going on.
+						var/damage = 5
+						var/obj/item/clothing/hat = attacker.head
+						if(istype(hat))
+							damage += hat.force * 3
+						affecting.apply_damage(damage*rand(90, 110)/100, BRUTE, "head", affected.run_armor_check(affecting, "melee"))
+						assailant.apply_damage(3*rand(90, 110)/100, BRUTE, "head", attacker.run_armor_check("head", "melee"))
```
Comment (Fox-McCloud) [inline code/modules/mob/mob_grab.dm]: Personally, I'd just remove the self damage now; 5 damage is low enough that there doesn't really need to be a downsides to headbutting, especially when the average unarmed melee damage is 4.5

**Comment by Fox-McCloud** [comment]
Everything seems to work great, though I couldn't get headbutting to work.

File: code/modules/mob/mob_grab.dm (hunk 1 above)
Comment (tigercat2000) [inline code/modules/mob/mob_grab.dm]: technically it could be as high as 15 damage if the hat had 3 force
